### PR TITLE
cockroachdb: 20.1.8 -> {20.1.17, 20.2.14, 21.1.7}

### PR DESCRIPTION
**Pull request by implr**
###### Motivation for this change
v21 is the current major stable version, v20.1 is EOL by upstream, but is kept here for the sake of clean upgrades.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jonringer** [comment]
> I'm somewhat anxious about nixpkgs skipping v20.2, as the upgrade documentation recommends against it.

We should probably have the major/minor versions that they recommend for an upgrade path available on nixpkgs

**Comment by implr** [comment]
Yeah ideally we'd have something like postgres, with multiple explicitly versioned packages. I didn't want to totally refactor this package, but I guess this has to be done.
BTW, cockroach <v21 officially doesn't support building with go >1.13, but that has been dropped from nixpkgs already. It does build though, they were just concerned with some timer performance issues.

**Comment by turion** [comment]
Can't we have `cockroach_20` which will be on the 20.2 version line?

**Comment by turion** [comment]
Or, rather introduce a `cockroach_21` and only bump `cockroach` to 20.2. Then when upstream announces EOL for 20.2, deprecate it, and then later move `cockroach` to 21.

**Comment by turion** [comment]
@implr do you want to continue working on this? Otherwise I'd pick it up.

**Comment by implr** [comment]
Oh, sorry, I didn't have time for this earlier. I pretty much have the packages written (we use them in a private repo), they just need to be adapted to nixpkgs. I should be able to do this later this week.

**Comment by implr** [comment]
Sorry for the delay.
`gopkg.nix` does the old-style buildGoPackage build for 20.1 - the package is unmodified from current master except for the .8 -> .17 bump. The other versions use the same approach as we have for postgres.

Another thing I'd like to do is add support for building the CCL licensed version (requires just changing the make target and some paths), but that can be a separate PR.

File: pkgs/servers/sql/cockroachdb/default.nix
Comment (turion) [inline pkgs/servers/sql/cockroachdb/default.nix:2]: How would you feel about removing this completely and only leave 20.2 and upwards?

Comment (turion) [reply pkgs/servers/sql/cockroachdb/default.nix:2]: This would have the advantage that it could be called `cockroachdb_20` and `cockroachdb_21`.

File: pkgs/servers/sql/cockroachdb/gopkg.nix
Comment (turion) [inline pkgs/servers/sql/cockroachdb/gopkg.nix:69]: Feel free to add me as a maintainer.

File: pkgs/top-level/all-packages.nix
Comment (turion) [inline pkgs/top-level/all-packages.nix:20835]: If we keep 20.1 I think there should be a deprecation warning.

**Comment by implr** [comment]
As far as naming goes, I think we need to keep two components, see https://www.cockroachlabs.com/docs/releases/index.html#release-naming
>A major release is produced twice a year indicating major enhancements to product functionality. A change in the YY.R component denotes a major release.

So a 21.2 when it comes it should be a separate package IMO to allow users to upgrade when they want.
Unsure about keeping 20.1. It *is* EOL, so it would need a deprecation warning and keeping it as a default is bad, but I don't know what to do about people who use it right now. What is the policy for postgres?

**Comment by turion** [comment]
`postgresql` has major versions 9 to 14 and then sets the "default" version to 13. Old versions (e.g. 9.6, released 5 years ago) get dropped as soon as they have no upstream maintenance anymore.

 `nextcloud` (server) on the other hand does not have a default version, it only has major versions and old (unmaintained) versions first get deprecated, then removed.

It's definitely the right approach to drop versions that are declared EOL/unmaintained upstream. Users need to stop using unmaintained software, and `nixpkgs` can't be burdened with keeping these versions around. If we offer them a range of supported versions such that they can go through the update process at their own pace. My personal opinion is that deprecation notices are an additional _service_ to users, giving them a grace period.

So we could mark 20.1.17 as deprecated on November 12th (https://www.cockroachlabs.com/docs/releases/v20.1.17) and remove it some time later.

File: pkgs/servers/sql/cockroachdb/default.nix
Comment (turion) [reply pkgs/servers/sql/cockroachdb/default.nix:2]: Ok, we've talked about this. We need the second component.

File: pkgs/top-level/all-packages.nix
Comment (turion) [reply pkgs/top-level/all-packages.nix]: 20.1 is still maintained though until November. I don't know whether we should add deprecation warnings _before_ a version gets unmaintained (with the date of the removal), or _after_. I think _before_ is nicer.

**Comment by turion** [comment]
> So we could mark 20.1.17 as deprecated on November 12th (https://www.cockroachlabs.com/docs/releases/v20.1.17) and remove it some time later.

Or rather mark it as deprecated _now_ and remove it on November 12th.

**Review by turion — CHANGES_REQUESTED** [review]
Just add deprecation notices, and then this MR looks perfect to me :)

**Comment by turion** [comment]
Ah, I thought about one more thing. Not sure about the commit message. Some people find it important to use specifically worded commit messages. Possibly it is better to have several commits, with messages:
```
cockroachdb: 20.1.8 -> 20.1.17
cockroachdb_20_2: init at 20.2.14
cockroachdb_21_1: init at 21.1.7
```
But I'm not sure whether that's worth splitting the commit up into three.

File: pkgs/servers/sql/cockroachdb/generic.nix
Comment (SuperSandro2000) [inline pkgs/servers/sql/cockroachdb/generic.nix:17]:
```suggestion
buildGoModule rec {
```

File: pkgs/servers/sql/cockroachdb/generic.nix
Comment (SuperSandro2000) [inline pkgs/servers/sql/cockroachdb/generic.nix:29]:
```suggestion
  nativeBuildInputs = [ installShellFiles cmake xz which autoconf ];

  buildInputs = if stdenv.isDarwin then [ libunwind libedit ] else [ ncurses6 ];
```

File: pkgs/servers/sql/cockroachdb/generic.nix
Comment (SuperSandro2000) [inline pkgs/servers/sql/cockroachdb/generic.nix:33]: This belongs into postPatch and should list which files/directories need patching instead of everything.

File: pkgs/servers/sql/cockroachdb/generic.nix
Comment (SuperSandro2000) [inline pkgs/servers/sql/cockroachdb/generic.nix:34]: Why can't we use makeFlags?

File: pkgs/servers/sql/cockroachdb/generic.nix
Comment (SuperSandro2000) [inline pkgs/servers/sql/cockroachdb/generic.nix:49]: This should be replaced with installManPage.

**Comment by turion** [comment]
@implr Ping, any progress? :)

File: pkgs/servers/sql/cockroachdb/generic.nix
Comment (implr) [reply pkgs/servers/sql/cockroachdb/generic.nix:34]: Cockroach's makefile is, uh, _cursed_ is probably the best word here really. eg. one thing it does is autodetect the number of cores and set various makeflags on its own, I'm not sure how that will interact with the standard nix machinery.

**Comment by implr** [comment]
Ah, I conflated maintenance support with assistance support, they're still providing security fixes.
> > So we could mark 20.1.17 as deprecated on November 12th (https://www.cockroachlabs.com/docs/releases/v20.1.17) and remove it some time later.
> 
> Or rather mark it as deprecated _now_ and remove it on November 12th.

Done.


> Ah, I thought about one more thing. Not sure about the commit message. Some people find it important to use specifically worded commit messages. Possibly it is better to have several commits, with messages:

Can do, but maybe as the very last thing (after all other comments are resolved).

@SuperSandro2000 re: all the style fixes - the package was like that before, I just left it as is, I guess I can change that.

File: pkgs/servers/sql/cockroachdb/default.nix
Comment (turion) [inline pkgs/servers/sql/cockroachdb/default.nix:2]:
```suggestion
  cockroachdb_20_1 = self.lib.warn ''
    Upstream support for Cockroach v20.1 ends on 2021-11-12.
    Please upgrade by using the `cockroachdb_20_2` or `cockroachdb_21_1` package,
    or prepare to be automatically updated as the old version will be removed after that date.
  '' (self.callPackage ./gopkg.nix {});
```

I guess we can make the message a bit more descriptive so that people know what to do

**Review by turion — COMMENTED** [review]
LGTM generally

File: pkgs/servers/sql/cockroachdb/default.nix
Comment (implr) [reply pkgs/servers/sql/cockroachdb/default.nix]: Is that sufficient?

Comment (turion) [reply pkgs/servers/sql/cockroachdb/default.nix]: Sounds good :+1:

**Comment by implr** [comment]
OK then, thanks. I need somebody else to merge though, don't have access.

**Comment by SuperSandro2000** [comment]
@ofborg eval

**Comment by SuperSandro2000** [comment]
First ofborg needs to be green before we can merge anything.

**Comment by implr** [comment]
Rebased. I'd like to finish this, has been dragging for a while.

Given the deprecation warning we agreed to above, ofborg will never be green, so unsure what to do with that.
Also it's an unfree package, does ofborg even eval those?

**Review by jonringer — APPROVED** [review]
overall diff LGTM

additional changes can be done in another PR. Waiting for ofborg to show green, then merge after building the packages.

File: pkgs/servers/sql/cockroachdb/gopkg.nix
Comment (SuperSandro2000) [inline pkgs/servers/sql/cockroachdb/gopkg.nix:61]: Can we add it to nativeBuildInputs?

File: pkgs/servers/sql/cockroachdb/gopkg.nix
Comment (SuperSandro2000) [inline pkgs/servers/sql/cockroachdb/gopkg.nix:51]: This can be done with install in one command.

File: pkgs/servers/sql/cockroachdb/gopkg.nix
Comment (SuperSandro2000) [inline pkgs/servers/sql/cockroachdb/gopkg.nix:17]: This can be done in the package with lib.optional quite nice.

**Comment by thequux** [comment]
Looking at the list of open review comments, it seems that they're all code style concerns. While keeping code quality high is important, I think that a bit of leniency on code style makes sense for this change; CockroachDB v20.1 will be deprecated before NixOS 21.11 gets released, and it would be best for the users who stay on stable to have a supported, if not maintained, version of CockroachDB available.

**Comment by ankhers** [comment]
Has there been any movement on this? I would be happy to try and fix things up as needed.

File: pkgs/servers/sql/cockroachdb/generic.nix
Comment (turion) [reply pkgs/servers/sql/cockroachdb/generic.nix:25]: How can we be more specific here?

Comment (turion) [reply pkgs/servers/sql/cockroachdb/generic.nix:25]: Ah, probably by finding all paths that need patching. Not sure why this is better. It will break quicker, but maybe take a little less build time.

File: pkgs/servers/sql/cockroachdb/default.nix
Comment (turion) [reply pkgs/servers/sql/cockroachdb/default.nix]: > Replace it with a throw alias.

This means that you need to add `cockroachdb_20_1` to `pkgs/top-level/aliases.nix`.

**Comment by turion** [comment]
20.1 is indeed already deprecated. To me it feels that this PR is actually great work, but has been slowed down with many code improvement details. It would be great to move forward soon so we can start maintaining the supported versions.

**Comment by turion** [comment]
> additional changes can be done in another PR. Waiting for ofborg to show green, then merge after building the packages.

@implr I think the only necessary change is to change the deprecation warning to an entry of `cockroachdb_20_1` to `pkgs/top-level/aliases.nix`. I'd merge after that.

@ankhers It would be great if you can add the code improvements after merging this PR :)

**Comment by implr** [comment]
> Has there been any movement on this? I would be happy to try and fix things up as needed.

Sorry everyone, I no longer have time to work on this, and I don't run cockroach at work anymore. Feel free to take over.

If you want to add the capability of building the commercial licensed versions, the changes I had in my package were adding `perl yacc` to deps and removing the `oss` suffix in build scripts.

**Comment by turion** [comment]
@ankhers fancy taking over? (Otherwise I can have a look middle of next year.)

File: pkgs/servers/sql/cockroachdb/default.nix
Comment (turion) [inline pkgs/servers/sql/cockroachdb/default.nix:16]: Looks like an extra newline?

**Comment by turion** [comment]
Closing in favour of https://github.com/NixOS/nixpkgs/pull/152626

**Pull request closed without merging**
